python/dgl/distributed/role.py

```diff
@@ -0,0 +1,200 @@
+    """Send client id and role to server
+
+    Parameters
+    ----------
+    client_id : int
+        ID of client
+    role : str
+        role of client
+    """
+    def __init__(self, client_id, machine_id, role):
+        self.client_id = client_id
+        self.machine_id = machine_id
+        self.role = role
+
+    def __getstate__(self):
+        return self.client_id, self.machine_id, self.role
+
+    def __setstate__(self, state):
+        self.client_id, self.machine_id, self.role = state
+
+    def process_request(self, server_state):
+        kv_store = server_state.kv_store
+        role = kv_store.role
+        if self.role not in role:
+            role[self.role] = set()
+            kv_store.barrier_count[self.role] = 0
+        role[self.role].add((self.client_id, self.machine_id))
+        total_count = 0
+        for key in role:
+            total_count += len(role[key])
+        # Clients are blocked util all clients register their roles.
+        if total_count == kv_store.num_clients:
+            res_list = []
+            for target_id in range(kv_store.num_clients):
+                print('send', target_id)
+                res_list.append((target_id, RegisterRoleResponse(REG_ROLE_MSG)))
+            return res_list
+        return None
+
+GET_ROLE = 700002
+GET_ROLE_MSG = "Get_Role"
+
+class GetRoleResponse(rpc.Response):
+    """Send the roles of all client processes"""
+    def __init__(self, role):
+        self.role = role
+        self.msg = GET_ROLE_MSG
+
+    def __getstate__(self):
+        return self.role, self.msg
+
+    def __setstate__(self, state):
+        self.role, self.msg = state
+
+class GetRoleRequest(rpc.Request):
+    """Send a request to get the roles of all client processes."""
+    def __init__(self):
+        self.msg = GET_ROLE_MSG
+
+    def __getstate__(self):
+        return self.msg
+
+    def __setstate__(self, state):
+        self.msg = state
+
+    def process_request(self, server_state):
+        kv_store = server_state.kv_store
+        return GetRoleResponse(kv_store.role)
+
+# The key is role, the value is a dict of mapping RPC rank to a rank within the role.
+PER_ROLE_RANK = {}
+
+# The global rank of a client process. The client processes of the same role have
+# global ranks that fall in a contiguous range.
+GLOBAL_RANK = {}
+
+# The role of the current process
+CUR_ROLE = None
+
+def init_role(role):
+    """Initialize the role of the current process.
+
+    Each process is associated with a role so that we can determine what
+    function can be invoked in a process. For example, we do not allow some
+    functions in sampler processes.
+
+    The initialization includes registeration the role of the current process and
+    get the roles of all client processes. It also computes the rank of all client
+    processes in a deterministic way so that all clients will have the same rank for
+    the same client process.
+    """
+    global CUR_ROLE
```

Comment:
Consiter put `CUR_ROLE` into `RPCContext` in` rpc.h `

Reply:
why?